chore: (qe) Preparatory refactoring to introduce log capturing.

miguelff opened the pull request:
Extracted from https://github.com/prisma/prisma-engines/pull/3505 and rebased.

This PR does a preparatory refactoring for the log capturing feature that can be reviewed and be released independently

Weakky commented on query-engine/request-handlers/src/lib.rs:
```diff
@@ -21,12 +19,3 @@ pub enum PrismaResponse {
     Multi(GQLBatchResponse),
 }
-
-impl PrismaResponse {
-    pub fn errors(&self) -> Vec<&GQLError> {
-        match self {
-            PrismaResponse::Single(ref s) => s.errors().collect(),
-            PrismaResponse::Multi(ref m) => m.errors().collect(),
-        }
-    }
-}
```
I suppose this was dead code?

miguelff replied:
Yes, there isn't any dynamic call to a particular prisma response

Weakky commented on query-engine/query-engine/src/server/mod.rs:
```diff
@@ -462,6 +356,37 @@ fn err_to_http_resp(err: query_core::CoreError) -> Response<Body> {
+#[allow(clippy::bind_instead_of_map)]
+fn transaction_id(headers: &HeaderMap) -> Option<TxId> {
+    const TRANSACTION_ID_HEADER: &str = "X-transaction-id";
+    headers
+        .get(TRANSACTION_ID_HEADER)
+        .and_then(|h| h.to_str().ok())
+        .and_then(|s| Some(TxId::from(s)))
```
nitpick

```suggestion
    headers
        .get(TRANSACTION_ID_HEADER)
        .and_then(|h| h.to_str().ok())
        .map(TxId::from)
```

Weakky reviewed (approved):
Great refactoring 💯. I left a few comments, I don't think any are blocking. Feel free to apply whichever makes the most sense to you.

tomhoule commented on query-engine/core/src/telemetry/helpers.rs:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    meta.target() == "quaint::connector::metrics" && meta.name() == "quaint:query"
+}
```
I see a lot of `pub fn` items in this file and elsewhere, and that is wholesale reexported as public API to consumers of the crate. Does this all need to be public? It make the API of a crate smaller and easier to understand if internal items are `pub(crate)` instead.

miguelff replied:
Right, I didn't restrict visibility of any of those, and just moved it. I will check and restrict when possible.

tomhoule commented on query-engine/core/src/telemetry/helpers.rs:
```diff
@@ -0,0 +1,112 @@
+    match serde_json::to_string(&span_result) {
+        Ok(json_string) => json_string,
+        Err(_) => "".to_string(),
+    }
```
Not a substantial improvement, but I think the whole `match` can be replaced with `serde_json::to_string(&span_result).unwrap_or_default()`.

tomhoule commented on query-engine/query-engine/src/context.rs:
```diff
@@ -43,7 +44,7 @@ impl ContextBuilder {
 impl PrismaContext {
     /// Initializes a new Prisma context.
-    async fn new(
+    pub async fn new(
```
I've searched the diff and found no reference to that function. Is making this public necessary because other code moved? (but that should be visible from the diff).

miguelff replied:
I splitted this PR from a different one, and that might required this function to be public. I will restrict and address if needed in the PR that adds capturing

tomhoule commented on query-engine/query-engine-node-api/src/logger.rs:
```diff
@@ -48,7 +48,7 @@ impl Logger {
 
         let log_callback_arc = Arc::new(log_callback);
-        let is_user_trace = filter_fn(is_user_facing_trace_filter);
+        let is_user_trace = filter_fn(telemetry::helpers::user_facing_span_only_filter);
```
:+1:

tomhoule commented on query-engine/query-engine/src/state.rs:
```diff
@@ -0,0 +1,84 @@
+
+impl State {
+    /// Create a new instance of `State`.
+    fn new(cx: PrismaContext, enable_playground: bool, enable_debug_mode: bool, enable_metrics: bool) -> Self {
```
No need to change this now, but boolean params are very poor in information when you read callers and really easy to get wrong (which boolean goes at which position). In my experience bitflags with enumflags are a lot easier to read when you construct them (all values are named), on top of being more compact in memory (here it would be 1 byte instead of 3). They're also more code and from a library, so, tradeoffs.

miguelff replied:
I totally agree, grabed the whole State object from a previous patch in progress, will try to address in here.

tomhoule replied:
Not blocking, like the rest of my feedback here, so do what you feel most comfortable with. No need to learn a whole (small) library for this if you haven't used `enumflags2` before.

miguelff replied:
I like the approach of using enum2flags for modeling a compacted set of flags. FWIW, these do already exist as boolean flags coming from the CLI. so we would have to make a transformation. Nevertheless, I don't want to add yet another boolean to this struct in case we have more flags, so I'm ok with exploring this option. Thanks for the suggestion.

I will do it, though, in a different PR after I merge both this and https://github.com/prisma/prisma-engines/pull/3618 to prevent more conflicts, as the https://github.com/prisma/prisma-engines/pull/3618 has changes using the state.

tomhoule commented on query-engine/query-engine/src/state.rs:
```diff
@@ -0,0 +1,84 @@
+            enable_metrics: self.enable_metrics,
+        }
+    }
+}
```
This looks identical to what `#[derive(Clone)]` on State would generate.

tomhoule commented on query-engine/query-engine/src/server/mod.rs:
```diff
@@ -462,6 +356,37 @@ fn err_to_http_resp(err: query_core::CoreError) -> Response<Body> {
+    match &value {
+        Some(str) if is_valid_traceparent(str) => value,
+        _ => None,
+    }
```
Some people really hate `Option::filter()` so don't feel forced to use it, but you can write that match statement as `value.filter(is_valid_traceparent)`

miguelff replied:
Why do people hate it, seems to make sense.

tomhoule replied:
My guess is years of JS experience where your brain automatically goes "oh, this is a collection/array" when you read "filter", and so you're surprised that it's a method on an Option.

miguelff replied:
OK, that's why I don't hate it lol

tomhoule commented on query-engine/core/src/telemetry/helpers.rs:
```diff
@@ -0,0 +1,112 @@
+use tracing_subscriber::EnvFilter;
+
+pub static SHOW_ALL_TRACES: Lazy<bool> = Lazy::new(|| match std::env::var("PRISMA_SHOW_ALL_TRACES") {
+    Ok(enabled) => enabled.to_lowercase() == *("true"),
```
I think that can be simplified to `enable.eq_ignore_ascii_case("true")`.

tomhoule commented on query-engine/core/src/executor/mod.rs:
```diff
@@ -53,20 +54,51 @@ pub trait QueryExecutor: TransactionManager {
+
+    /// Generates a new transaction id before the transaction is started and returns a modified version
+    /// of self with the new predefined_id set.
+    pub fn with_new_transaction_id(&mut self) -> TxId {
+        let tx_id: TxId = Default::default();
+        self.new_tx_id = Some(tx_id.to_string());
+        tx_id
+    }
```
Yes, you can even have `#[serde(deserialize_with)]` on the field, which would make the logic private to this module, so harder to get wrong in other contexts.

miguelff replied:
Absolutely, good call! will change

tomhoule commented on query-engine/query-engine/src/logger.rs:
```diff
@@ -51,40 +55,34 @@ impl<'a> Logger<'a> {
-    pub fn telemetry_endpoint(&mut self, endpoint: &'a str) {
-        if endpoint.is_empty() {
-            self.telemetry_endpoint = None
+    pub fn setup_telemetry(&mut self, enable_telemetry: bool, endpoint: &str) {
```
I am always wondering with these boolean parameters: would it make sense to have `if enable_telemetry { setup_telemetry(endpoint) }` in the callers rather than this? I am likely missing something subtle.

tomhoule commented on query-engine/query-engine/src/logger.rs:
```diff
@@ -1,43 +1,47 @@
+#[derive(Debug, Clone)]
+enum TracingConfig {
+    // exposed means tracing will be exposed through an HTTP endpoint in a jaeger-compatible format
+    Exposed(String),
```
Is the string data the endpoint?

Owned strings look easier on the surface but they open other challenges with the borrow checker, and often harder ones — was it made owned for convenience?

miguelff replied:
Yes, it only was made owned for convenience. I can introduce a lifetime, and see how it looks like.

tomhoule commented on query-engine/query-engine/src/logger.rs:
```diff
@@ -99,54 +97,39 @@ impl<'a> Logger<'a> {
+            TracingConfig::Disabled => {
+                subscriber::set_global_default(subscriber)?;
+            }
+        }
```
I like this :+1:

tomhoule commented on query-engine/query-engine/src/state.rs:
```diff
@@ -0,0 +1,84 @@
+    let enable_metrics = config.preview_features().contains(PreviewFeature::Metrics) || opts.dataproxy_metric_override;
+    let span = tracing::info_span!("prisma:engine:connect");
+
+    let cx = PrismaContext::builder(datamodel) //  opts.enable_raw_queries, metrics, logs_capture)
```
I think the trailing comment may not be intended.

miguelff replied:
You are right

tomhoule reviewed (approved):
Mostly superficial comments on details rather than the big picture. I am still lacking a lot of context to feel onboarded on tracing in the QE, and I have to learn opentelemetry (it's a lot more than a small layer on top of `tracing`).

tomhoule commented:
(none of my feedback is blocking)

miguelff commented on query-engine/query-engine/src/server/mod.rs:
```diff
@@ -121,15 +39,15 @@ pub async fn listen(opts: PrismaOpt, metrics: MetricRegistry) -> PrismaResult<()
     {
-        return handle_metrics(state, req).await;
+        return metrics_handler(state, req).await;
     }
```
Yeah, me too!

miguelff commented on query-engine/query-engine/src/logger.rs:
```diff
@@ -1,43 +1,47 @@
+#[derive(Debug, Clone)]
+enum TracingConfig {
+    // exposed means tracing will be exposed through an HTTP endpoint in a jaeger-compatible format
+    Exposed(String),
```
I like it, yes!

miguelff commented:
Thank you both @Weakky and @tomhoule for the sensible feedback, it also gives me a lot of good information that I didn't know about and that would help improve the patch, which is the best outcome possible for a review. Thank you so much! 🤗 

Will follow up with a new commit addressing the feedback and merge if all good.

miguelff commented on query-engine/core/src/interactive_transactions/mod.rs:
```diff
@@ -54,7 +56,15 @@ where
     fn from(s: T) -> Self {
-        Self(s.into())
+        let contents = s.into();
+        assert!(
+            contents.len() >= MINIMUM_TX_ID_LENGTH,
+            "minimum length for a TxId ({}) is {}, but was {}",
+            contents,
+            MINIMUM_TX_ID_LENGTH,
+            contents.len()
+        );
```
This is to guarantee that we can derive a TraceId from a TxId, by ensuring the TxId has the format expected (this is a relaxed precondition over the format of a TxId, the own TxId module doesn't have an API exposed to validate them, nor their underlying `cuid` module )

miguelff replied:
Will document
